client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java

```diff
@@ -0,0 +1,157 @@
+    private final Function<String, Future<DelegatedCommandSender>> delegatedCommandSenderSupplier;
+
+    /**
+     * Creates a new DelegatingCommandHandler.
+     *
+     * @param deviceSpecificCommandHandlerSupplier Function to get an existing command handler for a given device.
+     * @param delegatedCommandSenderSupplier Supplies a sender for a given tenant to send the delegated command via the
+     *            downstream peer.
+     */
+    public DelegatingCommandHandler(final Function<String, Handler<CommandContext>> deviceSpecificCommandHandlerSupplier,
+                                    final Function<String, Future<DelegatedCommandSender>> delegatedCommandSenderSupplier) {
+        this.deviceSpecificCommandHandlerSupplier = deviceSpecificCommandHandlerSupplier;
+        this.delegatedCommandSenderSupplier = delegatedCommandSenderSupplier;
+    }
+
+
+    @Override
+    public void handle(final CommandContext commandContext) {
+        final String deviceId = commandContext.getCommand().getDeviceId();
+        final Handler<CommandContext> commandHandler = deviceSpecificCommandHandlerSupplier.apply(deviceId);
+        if (commandHandler != null) {
+            // delegate to existing local device-specific handler
+            LOG.trace("use existing command handler for device {}", deviceId);
+            commandHandler.handle(commandContext);
+        } else {
+            // delegate to matching consumer via downstream peer
+            LOG.trace("delegate command for device {} to matching consumer via downstream peer", deviceId);
+            delegateReceivedCommandMessageViaDownstreamPeer(commandContext);
+        }
+    }
+
+    private void delegateReceivedCommandMessageViaDownstreamPeer(final CommandContext commandContext) {
+
+        final Command command = commandContext.getCommand();
+        final String tenantId = command.getTenant();
+        final String deviceId = command.getDeviceId();
+        // send message to AMQP network
+        final Future<DelegatedCommandSender> delegatedCommandSender = delegatedCommandSenderSupplier.apply(tenantId);
+        delegatedCommandSender.setHandler(cmdSenderResult -> {
+            if (cmdSenderResult.succeeded()) {
+                final MessageSender sender = cmdSenderResult.result();
+                final String targetAddress = DelegatedCommandSenderImpl.getTargetAddress(tenantId, deviceId);
+                final String replyToAddress = command.isOneWay() ? null
+                        : String.format("%s/%s/%s",
+                                CommandConstants.COMMAND_ENDPOINT, tenantId, command.getReplyToId());
+                final Future<ProtonDelivery> protonDeliveryFuture = sender.sendAndWaitForOutcome(
+                        createDelegatedCommandMessage(command.getCommandMessage(), targetAddress, replyToAddress),
+                        commandContext.getTracingContext());
```

Comment:
Isn't this the specific command context handling that `DelegatedCommandSender` could/should encapsulate? Why don't we define a method `DelegatedCommandSender.send(CommandContext)`?

Reply:
Yes, added a method in the interface (see above).